### ISSUE TITLE
Unexpected / new "error: value is a path while a set was expected" when importing "nixos-hardware.nixosModules.asus-zephyrus-ga402x.amdgpu" or "nixos-hardware.nixosModules.asus-zephyrus-ga402x.nvidia"

### ISSUE THREAD

**Issue by mexisme** [NixOS/nixos-hardware#1052]
I think PR #1046 has broken my importing of `nixos-hardware.nixosModules.asus-zephyrus-ga402x.amdgpu`, which was working beforehand, but now gives the error:
```
error: value is a path while a set was expected
```

This really confused me until (after some digging) I realised that `nixos-hardware.nixosModules.asus-zephyrus-ga402x` now pointed to a separate source path, rather than a direct import of the `default.nix` at [`//asus/zephyrus/ga402x/default.nix`](https://github.com/NixOS/nixos-hardware/blob/master/asus/zephyrus/ga402x/default.nix), which does export a `amdgpu` and `nvidia` attr.

This made it easy to switch between AMD only and NVidia boot specialisations, because disabling NVidia completely saved some battery — let alone for those times where the AMD GPU driver _or_ the NVidia driver have a regression.

NOTE:
Also breaks `lenovo-yoga-7-14ARH7.amdgpu` and `lenovo-yoga-7-14ARH7.nvidia`

**Comment by mexisme** [comment]
// @NiklasGollenstede

**Comment by mexisme** [comment]
I suspect it's not useful, but the full error message (for me) looks like:
```
error:
       … while calling the 'seq' builtin

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:322:18:

          321|         options = checked options;
          322|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          323|         _module = checked (config._module);

       … while evaluating a branch condition

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:261:9:

          260|       checkUnmatched =
          261|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |         ^
          262|           let

       … in the left operand of the AND (&&) operator

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:261:72:

          260|       checkUnmatched =
          261|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |                                                                        ^
          262|           let

       … in the left operand of the AND (&&) operator

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:261:33:

          260|       checkUnmatched =
          261|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |                                 ^
          262|           let

       … while evaluating a branch condition

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:254:12:

          253|
          254|         in if declaredConfig._module.freeformType == null then declaredConfig
             |            ^
          255|           # Because all definitions that had an associated option ended in

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:242:28:

          241|           # For definitions that have an associated option
          242|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                            ^
          243|

       … while calling 'mapAttrsRecursiveCond'

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/attrsets.nix:1201:5:

         1200|     f:
         1201|     set:
             |     ^
         1202|     let

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:234:33:

          233|           ({ inherit lib options config specialArgs; } // specialArgs);
          234|         in mergeModules prefix (reverseList collected);
             |                                 ^
          235|

       … while calling 'reverseList'

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/lists.nix:1116:17:

         1115|   */
         1116|   reverseList = xs:
             |                 ^
         1117|     let l = length xs; in genList (n: elemAt xs (l - n - 1)) l;

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:229:25:

          228|       merged =
          229|         let collected = collectModules
             |                         ^
          230|           class

       … while calling anonymous lambda

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:445:37:

          444|
          445|     in modulesPath: initialModules: args:
             |                                     ^
          446|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:446:7:

          445|     in modulesPath: initialModules: args:
          446|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);
             |       ^
          447|

       … while calling 'filterModules'

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:413:36:

          412|       # modules recursively. It returns the final list of unique-by-key modules
          413|       filterModules = modulesPath: { disabled, modules }:
             |                                    ^
          414|         let

       … while calling anonymous lambda

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:439:31:

          438|           disabledKeys = concatMap ({ file, disabled }: map (moduleKey file) disabled) disabled;
          439|           keyFilter = filter (attrs: ! elem attrs.key disabledKeys);
             |                               ^
          440|         in map (attrs: attrs.module) (builtins.genericClosure {

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:400:22:

          399|           let
          400|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
             |                      ^
          401|             collectedImports = collectStructuredModules module._file module.key module.imports args;

       … while calling anonymous lambda

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:359:11:

          358|         then
          359|           m:
             |           ^
          360|             if m._class != null -> m._class == class

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:400:35:

          399|           let
          400|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
             |                                   ^
          401|             collectedImports = collectStructuredModules module._file module.key module.imports args;

       … while calling 'loadModule'

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:336:53:

          335|       # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
          336|       loadModule = args: fallbackFile: fallbackKey: m:
             |                                                     ^
          337|         if isFunction m then

       … from call site

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/modules.nix:337:12:

          336|       loadModule = args: fallbackFile: fallbackKey: m:
          337|         if isFunction m then
             |            ^
          338|           unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgs fallbackKey m args)

       … while calling 'isFunction'

         at /nix/store/qszplw617r895nbcprgyj139c9a3r0xs-source/lib/trivial.nix:929:16:

          928|   */
          929|   isFunction = f: builtins.isFunction f ||
             |                ^
          930|     (f ? __functor && isFunction (f.__functor f));

       error: value is a path while a set was expected
```

**Comment by nipsy** [comment]
Just ran into this last night when defining nixos-hardware.nixosModules.lenovo-yoga-7-14ARH7.amdgpu in my module list for my laptop's nixosSystem.  Since I had other NixOS systems building without any errors and the inclusion of this hardware flake was pretty much the only difference for that system, I chanced upon the correct solution by commenting out that line in my module list and then found this bug report after the rebuild succeeded that time.

The error output certainly didn't help to narrow it down initially.  The initial error was after the first failure where the extra eval-cache option was necessary now to attempt a rebuild:

```
# nixos-rebuild switch --show-trace --flake .#
error: cached failure of attribute 'nixosConfigurations.ginaz.config'

# nixos-rebuild switch --option eval-cache false --show-trace --flake .#
error:
       … while calling the 'seq' builtin

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:334:18:

          333|         options = checked options;
          334|         config = checked (removeAttrs config [ "_module" ]);
             |                  ^
          335|         _module = checked (config._module);

       … while evaluating a branch condition

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:9:

          272|       checkUnmatched =
          273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |         ^
          274|           let

       … in the left operand of the AND (&&) operator

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:72:

          272|       checkUnmatched =
          273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |                                                                        ^
          274|           let

       … in the left operand of the AND (&&) operator

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:33:

          272|       checkUnmatched =
          273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
             |                                 ^
          274|           let

       … while evaluating a branch condition

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:266:12:

          265|
          266|         in if declaredConfig._module.freeformType == null then declaredConfig
             |            ^
          267|           # Because all definitions that had an associated option ended in

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:254:28:

          253|           # For definitions that have an associated option
          254|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                            ^
          255|

       … while calling 'mapAttrsRecursiveCond'

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/attrsets.nix:1201:5:

         1200|     f:
         1201|     set:
             |     ^
         1202|     let

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:246:33:

          245|           ({ inherit lib options config specialArgs; } // specialArgs);
          246|         in mergeModules prefix (reverseList collected);
             |                                 ^
          247|

       … while calling 'reverseList'

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/lists.nix:1116:17:

         1115|   */
         1116|   reverseList = xs:
             |                 ^
         1117|     let l = length xs; in genList (n: elemAt xs (l - n - 1)) l;

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:241:25:

          240|       merged =
          241|         let collected = collectModules
             |                         ^
          242|           class

       … while calling anonymous lambda

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:457:37:

          456|
          457|     in modulesPath: initialModules: args:
             |                                     ^
          458|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:458:7:

          457|     in modulesPath: initialModules: args:
          458|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);
             |       ^
          459|

       … while calling 'filterModules'

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:425:36:

          424|       # modules recursively. It returns the final list of unique-by-key modules
          425|       filterModules = modulesPath: { disabled, modules }:
             |                                    ^
          426|         let

       … while calling anonymous lambda

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:451:31:

          450|           disabledKeys = concatMap ({ file, disabled }: map (moduleKey file) disabled) disabled;
          451|           keyFilter = filter (attrs: ! elem attrs.key disabledKeys);
             |                               ^
          452|         in map (attrs: attrs.module) (genericClosure {

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:412:22:

          411|           let
          412|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
             |                      ^
          413|             collectedImports = collectStructuredModules module._file module.key module.imports args;

       … while calling anonymous lambda

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:371:11:

          370|         then
          371|           m:
             |           ^
          372|             if m._class != null -> m._class == class

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:412:35:

          411|           let
          412|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
             |                                   ^
          413|             collectedImports = collectStructuredModules module._file module.key module.imports args;

       … while calling 'loadModule'

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:348:53:

          347|       # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
          348|       loadModule = args: fallbackFile: fallbackKey: m:
             |                                                     ^
          349|         if isFunction m then

       … from call site

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:349:12:

          348|       loadModule = args: fallbackFile: fallbackKey: m:
          349|         if isFunction m then
             |            ^
          350|           unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgs fallbackKey m args)

       … while calling 'isFunction'

         at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/trivial.nix:989:16:

          988|   */
          989|   isFunction = f: builtins.isFunction f ||
             |                ^
          990|     (f ? __functor && isFunction (f.__functor f));

       error: value is a path while a set was expected
```

This is one of my few complaints with NixOS is that the error messages are next to useless sometimes in figuring out where the problem actually lies.  Although I'm open to suggestions if a better command line invocation would have somehow sussed this out as a problem in this particular flake.

**Comment by mexisme** [comment]
> Just ran into this last night when defining nixos-hardware.nixosModules.lenovo-yoga-7-14ARH7.amdgpu in my module list for my laptop's nixosSystem. Since I had other NixOS systems building without any errors and the inclusion of this hardware flake was pretty much the only difference for that system, I chanced upon the correct solution by commenting out that line in my module list and then found this bug report after the rebuild succeeded that time.
> 
> The error output certainly didn't help to narrow it down initially. The initial error was after the first failure where the extra eval-cache option was necessary now to attempt a rebuild:
> 
> ```
> # nixos-rebuild switch --show-trace --flake .#
> error: cached failure of attribute 'nixosConfigurations.ginaz.config'
> 
> # nixos-rebuild switch --option eval-cache false --show-trace --flake .#
> error:
>        … while calling the 'seq' builtin
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:334:18:
> 
>           333|         options = checked options;
>           334|         config = checked (removeAttrs config [ "_module" ]);
>              |                  ^
>           335|         _module = checked (config._module);
> 
>        … while evaluating a branch condition
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:9:
> 
>           272|       checkUnmatched =
>           273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
>              |         ^
>           274|           let
> 
>        … in the left operand of the AND (&&) operator
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:72:
> 
>           272|       checkUnmatched =
>           273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
>              |                                                                        ^
>           274|           let
> 
>        … in the left operand of the AND (&&) operator
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:273:33:
> 
>           272|       checkUnmatched =
>           273|         if config._module.check && config._module.freeformType == null && merged.unmatchedDefns != [] then
>              |                                 ^
>           274|           let
> 
>        … while evaluating a branch condition
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:266:12:
> 
>           265|
>           266|         in if declaredConfig._module.freeformType == null then declaredConfig
>              |            ^
>           267|           # Because all definitions that had an associated option ended in
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:254:28:
> 
>           253|           # For definitions that have an associated option
>           254|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
>              |                            ^
>           255|
> 
>        … while calling 'mapAttrsRecursiveCond'
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/attrsets.nix:1201:5:
> 
>          1200|     f:
>          1201|     set:
>              |     ^
>          1202|     let
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:246:33:
> 
>           245|           ({ inherit lib options config specialArgs; } // specialArgs);
>           246|         in mergeModules prefix (reverseList collected);
>              |                                 ^
>           247|
> 
>        … while calling 'reverseList'
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/lists.nix:1116:17:
> 
>          1115|   */
>          1116|   reverseList = xs:
>              |                 ^
>          1117|     let l = length xs; in genList (n: elemAt xs (l - n - 1)) l;
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:241:25:
> 
>           240|       merged =
>           241|         let collected = collectModules
>              |                         ^
>           242|           class
> 
>        … while calling anonymous lambda
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:457:37:
> 
>           456|
>           457|     in modulesPath: initialModules: args:
>              |                                     ^
>           458|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:458:7:
> 
>           457|     in modulesPath: initialModules: args:
>           458|       filterModules modulesPath (collectStructuredModules unknownModule "" initialModules args);
>              |       ^
>           459|
> 
>        … while calling 'filterModules'
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:425:36:
> 
>           424|       # modules recursively. It returns the final list of unique-by-key modules
>           425|       filterModules = modulesPath: { disabled, modules }:
>              |                                    ^
>           426|         let
> 
>        … while calling anonymous lambda
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:451:31:
> 
>           450|           disabledKeys = concatMap ({ file, disabled }: map (moduleKey file) disabled) disabled;
>           451|           keyFilter = filter (attrs: ! elem attrs.key disabledKeys);
>              |                               ^
>           452|         in map (attrs: attrs.module) (genericClosure {
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:412:22:
> 
>           411|           let
>           412|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
>              |                      ^
>           413|             collectedImports = collectStructuredModules module._file module.key module.imports args;
> 
>        … while calling anonymous lambda
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:371:11:
> 
>           370|         then
>           371|           m:
>              |           ^
>           372|             if m._class != null -> m._class == class
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:412:35:
> 
>           411|           let
>           412|             module = checkModule (loadModule args parentFile "${parentKey}:anon-${toString n}" x);
>              |                                   ^
>           413|             collectedImports = collectStructuredModules module._file module.key module.imports args;
> 
>        … while calling 'loadModule'
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:348:53:
> 
>           347|       # Like unifyModuleSyntax, but also imports paths and calls functions if necessary
>           348|       loadModule = args: fallbackFile: fallbackKey: m:
>              |                                                     ^
>           349|         if isFunction m then
> 
>        … from call site
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/modules.nix:349:12:
> 
>           348|       loadModule = args: fallbackFile: fallbackKey: m:
>           349|         if isFunction m then
>              |            ^
>           350|           unifyModuleSyntax fallbackFile fallbackKey (applyModuleArgs fallbackKey m args)
> 
>        … while calling 'isFunction'
> 
>          at /nix/store/afawjm6nv8m5aagkcqxlxgrr5qpyh6cm-source/lib/trivial.nix:989:16:
> 
>           988|   */
>           989|   isFunction = f: builtins.isFunction f ||
>              |                ^
>           990|     (f ? __functor && isFunction (f.__functor f));
> 
>        error: value is a path while a set was expected
> ```
> 
> This is one of my few complaints with NixOS is that the error messages are next to useless sometimes in figuring out where the problem actually lies. Although I'm open to suggestions if a better command line incantation would have somehow sussed this out as a problem in this particular flake.

TL;DR:
I think the only places where the (incorrect) attr-set style was used was in `asus-zephyrus-ga402x` and `lenovo-yoga-7-14ARH7` — both of which were my fault, I'm afraid.

---

IMHO, in this particular situation, it's more that the inner complexities of modules can make it harder to diagnose higher-level errors like this.

As I understand it, part of the reason for @NiklasGollenstede's change was to try to make it easier to diagnose module errors;  unfortunately this particular one is much less obvious because of when it fails.
I only realised it because of a bit of an epiphany!
